javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -658,36 +622,45 @@ public Double getNext() {
         } else if (step == 0) {
             throw new IllegalArgumentException("step cannot be 0");
         } else if (from == toInclusive) {
-            return Iterator.of(from);
-        } else if (step * (from - toInclusive) > 0) {
-            return Iterator.empty();
+            return of(from);
+        } else if (Math.signum(step) == Math.signum(from - toInclusive)) {
+            return empty();
         } else {
-            return new AbstractIterator<Double>() {
-
-                double prev = Double.NaN;
-                double curr = from;
-                boolean hasNext = true;
+            if (step > 0) {
+                return new AbstractIterator<Double>() {
+                    final double end = Math.nextUp(toInclusive);
+                    double i = from;
+
+                    @Override
+                    public boolean hasNext() {
+                        return i < end;
+                    }
 
-                @Override
-                public boolean hasNext() {
-                    return hasNext;
-                }
+                    @Override
+                    public Double getNext() {
+                        final double next = i;
+                        i = Math.max(i + step, Math.nextUp(i));
```

Comment:
When you predict the price of your assets for the next 30 years it is ok to have an error. Nobody knows the future :)

When you calculate those prices for your whole portfolio, let's say for hundreds of thousands of trades, and you have to calculate scenarios and to shift values for hundreds of days back in time, then the calculation has to be *really* fast. We have 90 nodes in our calculation grid and the net time is 250 hours calculation when calculating the risk for example!